src/go/k8s/internal/controller/pvcunbinder/pvcunbinder.go

```diff
@@ -0,0 +1,314 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// Reconciler is a Kubernetes Reconciler that watches for Pods stuck in a
+// Pending state due to volume affinities and attempts a remediation.
+//
+// It watches for Pod events rather than Node events because:
+//  1. Node Deletion events could be missed if the operator is scheduled on the node that's died
+//  2. We don't want to re-implement label matching. In theory, it should be
+//     easy but it's quite risky and behaviors could diverge between Kubernetes
+//     versions.
+//
+// To get the Pod to reschedule we:
+//  1. Find all PVs and PVCs associated with our Pod.
+//  2. Ensure that all PVs in question have a Retain policy
+//  3. Delete all PVCs from step 1. (PVCs are immutable after creation,
+//     deletion is the only option)
+//  4. "Recycle" all PVs from step 1 by clearing the ClaimRef.
+//     Kubernetes will only consider binding PVs that have a satisfiable
+//     NodeAffinity. By "recycling" we permit Flakey Nodes to rejoin the cluster
+//     which _might_ reclaim the now freed volume.
+//  5. Deleting the Pod to re-trigger PVC creation and rebinding.
+type Reconciler struct {
+	Client client.Client
+	// Timeout is the duration a Pod must be stuck in Pending before
+	// remediation is attempted.
+	Timeout time.Duration
+	// Selector, if specified, will narrow the scope of Pods that this
+	// Reconciler will consider for remediation.
+	Selector labels.Selector
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).For(&corev1.Pod{}, builder.WithPredicates(predicate.NewPredicateFuncs(pvcUnbinderPredicate))).Complete(r)
+}
+
+// Reconcile implements the algorithm described in the docs of [Reconciler]. To
+// the best of it's ability, Reconcile is implemented to be idempotent. Due to
+// the lack of transactions in Kubernetes/etc and the need to operate across
+// many objects, it's quite difficult to guarantee this. The general strategy
+// is to fetch a snapshot of the world as early as possible and then rely on
+// ResourceVersions to inform us about changes from external actors, in which
+// case we'll re-queue.
+// TODO use an in memory timeout to prevent complete unbinding of Pods.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx).WithName("PVCUnbinder")
+	ctx = log.IntoContext(ctx, logger)
+
+	var pod corev1.Pod
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if ok, requeueAfter := r.shouldRemediate(ctx, &pod); !ok || requeueAfter > 0 {
+		logger.Info("shouldn't remediate Pod; skipping", "name", pod.Name, "ok", ok, "requeue-after", requeueAfter)
+		return ctrl.Result{RequeueAfter: requeueAfter, Requeue: ok}, nil
+	}
+
+	// NB: We denote PVCs that are deleted as a nil entry within this map. If a
```

Comment:
Nit: you could encode this logic by hiding the map within a struct as a private field and exposing `MarkDeleted(key client.ObjectKey)` and `Ignore(key client.ObjectKey)`.

Reply:
Good call out! I'm going to decline making this change unless this package/function becomes more complex. The extra code volume and loss of map ergonomics doesn't currently feel like a worthwhile trade off. Once/if iterators land in go 1.23, this would be a very nice change.